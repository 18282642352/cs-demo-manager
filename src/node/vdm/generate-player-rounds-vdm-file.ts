import { VDMGenerator } from './generator';
import type { Round } from '../counter-strike/launcher/watch-player-rounds';

type Options = {
  demoPath: string;
  rounds: Round[];
  steamId: string;
};

export async function generatePlayerRoundsVdmFile({ demoPath, rounds, steamId }: Options) {
  const vdm = new VDMGenerator(demoPath);

  let currentTick = 0;
  for (let index = 0; index < rounds.length; index++) {
    const round = rounds[index];
    vdm.addSkipAhead(currentTick, round.freezeTimeEndTick);
    vdm.addSpecPlayer(currentTick, steamId);

    const roundEndTick = rounds[index]?.tickEnd ?? 0;
    currentTick = roundEndTick + 128;

    if (index === rounds.length - 1) {
      vdm.addStopPlayback(currentTick);
    }
  }

  await vdm.write();

  return vdm;
}
