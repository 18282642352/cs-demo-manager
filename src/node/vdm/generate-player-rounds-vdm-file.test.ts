import { fileURLToPath } from 'node:url';
import fs from 'node:fs/promises';
import { describe, expect, it } from 'vitest';
import { generatePlayerRoundsVdmFile } from './generate-player-rounds-vdm-file';
import type { Round } from '../counter-strike/launcher/watch-player-rounds';

describe(`Generate player's rounds VDM file`, () => {
  it(`should generate a VDM file that to watch player's rounds`, async () => {
    const demoPath = fileURLToPath(new URL('player-rounds.dem', import.meta.url));
    const fakePlayerSteamId = 'fakeKillerSteamId';
    const rounds: Round[] = [
      {
        number: 1,
        freezeTimeEndTick: 10,
        tickEnd: 90,
      },
      {
        number: 2,
        freezeTimeEndTick: 110,
        tickEnd: 190,
      },
      {
        number: 3,
        freezeTimeEndTick: 210,
        tickEnd: 290,
      },
      {
        number: 4,
        freezeTimeEndTick: 310,
        tickEnd: 390,
      },
    ];

    const vdm = await generatePlayerRoundsVdmFile({
      demoPath,
      rounds,
      steamId: fakePlayerSteamId,
    });

    const vdmContent = await fs.readFile(vdm.getVdmPath(), 'utf8');
    expect(vdmContent).toMatchSnapshot();
  });
});
