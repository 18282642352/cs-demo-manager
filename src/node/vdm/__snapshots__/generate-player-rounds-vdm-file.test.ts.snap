// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generate player's rounds VDM file > should generate a VDM file that to watch player's rounds 1`] = `
"
// Generated by CS Demo Manager
demoactions
{
  
  "1"
  {
    factory "SkipAhead"
    name "SkipAhead1"
    starttick "0"
    skiptotick "10"
  }
  "2"
  {
    factory "PlayCommands"
    name "spec2"
    starttick "0"
    commands "spec_lock_to_accountid fakeKillerSteamId"
  }
  "3"
  {
    factory "PlayCommands"
    name "spec3"
    starttick "0"
    commands "spec_player_by_accountid fakeKillerSteamId"
  }
  "4"
  {
    factory "SkipAhead"
    name "SkipAhead4"
    starttick "218"
    skiptotick "110"
  }
  "5"
  {
    factory "PlayCommands"
    name "spec5"
    starttick "218"
    commands "spec_lock_to_accountid fakeKillerSteamId"
  }
  "6"
  {
    factory "PlayCommands"
    name "spec6"
    starttick "218"
    commands "spec_player_by_accountid fakeKillerSteamId"
  }
  "7"
  {
    factory "SkipAhead"
    name "SkipAhead7"
    starttick "268"
    skiptotick "210"
  }
  "8"
  {
    factory "PlayCommands"
    name "spec8"
    starttick "268"
    commands "spec_lock_to_accountid fakeKillerSteamId"
  }
  "9"
  {
    factory "PlayCommands"
    name "spec9"
    starttick "268"
    commands "spec_player_by_accountid fakeKillerSteamId"
  }
  "10"
  {
    factory "SkipAhead"
    name "SkipAhead10"
    starttick "358"
    skiptotick "310"
  }
  "11"
  {
    factory "PlayCommands"
    name "spec11"
    starttick "358"
    commands "spec_lock_to_accountid fakeKillerSteamId"
  }
  "12"
  {
    factory "PlayCommands"
    name "spec12"
    starttick "358"
    commands "spec_player_by_accountid fakeKillerSteamId"
  }
  "13"
  {
    factory "StopPlayback"
    name "stop13"
    starttick "518"
  }
}
"
`;
